File: go-probe/resultstore.go
// Хранилище результатов с подтверждением доставки (ACK) — аналог C# ResultStore.
//
// Результаты копятся в очереди; сервер забирает их пачкой (TakeBatch), пачка
// «в полёте» хранится до подтверждения (Confirm) и при потере связи выдаётся
// повторно — дубликаты сервер отбрасывает по ResultId. Очередь ограничена:
// при переполнении вытесняются самые старые. Снимок недоставленного пишется
// на диск по таймеру (не на каждый результат) — дисковая нагрузка минимальна.
package main

import (
	"encoding/json"
	"os"
	"slices"
	"sync"
	"time"
)

// resultsFileName — файл недоставленных результатов между перезапусками.
const resultsFileName = "JobResult.json"

// ResultStore — очередь результатов с подтверждением и сохранением на диск.
type ResultStore struct {
	mu         sync.Mutex
	pending    []ActionData
	inFlightId string
	inFlight   []ActionData
	maxPending int
	dropped    int64

	signal  chan struct{} // сигнал «появились данные» для длинного опроса
	dirty   bool
	persist time.Duration
	stop    chan struct{}
}

// NewResultStore создаёт хранилище и запускает фоновый цикл сохранения.
func NewResultStore(maxPending, persistSec int) *ResultStore {
	s := &ResultStore{
		maxPending: maxPending,
		signal:     make(chan struct{}, 1),
		persist:    time.Duration(persistSec) * time.Second,
		stop:       make(chan struct{}),
	}
	go s.persistLoop()
	return s
}

// Add ставит результат в очередь (с вытеснением самых старых при переполнении).
func (s *ResultStore) Add(result ActionData) {
	s.mu.Lock()
	s.pending = append(s.pending, result)
	if len(s.pending) > s.maxPending {
		s.pending = s.pending[1:]
		s.dropped++
		if s.dropped == 1 || s.dropped%1000 == 0 {
			logResults.Error("Очередь результатов переполнена — старые записи вытесняются",
				"лимит", s.maxPending, "всего_вытеснено", s.dropped)
		}
	}
	s.dirty = true
	s.mu.Unlock()

	select {
	case s.signal <- struct{}{}:
	default: // сигнал уже стоит — потребитель и так проснётся
	}
}

// TakeBatch выдаёт пачку результатов, ожидая появления данных до timeout
// («длинный опрос»). Неподтверждённая пачка выдаётся повторно.
func (s *ResultStore) TakeBatch(timeout time.Duration) ResultBatch {
	deadline := time.After(timeout)
	for {
		s.mu.Lock()
		if s.inFlight != nil {
			batch := ResultBatch{BatchId: s.inFlightId, Items: slices.Clone(s.inFlight)} // slices.Clone — Go 1.21
			s.mu.Unlock()
			return batch
		}
		if len(s.pending) > 0 {
			s.inFlight = s.pending
			s.pending = nil
			s.inFlightId = NewGuid()
			batch := ResultBatch{BatchId: s.inFlightId, Items: slices.Clone(s.inFlight)}
			s.mu.Unlock()
			return batch
		}
		s.mu.Unlock()

		select {
		case <-s.signal: // появились данные — попробуем снова
		case <-deadline:
			return ResultBatch{BatchId: EmptyGuid, Items: []ActionData{}}
		}
	}
}

// Confirm подтверждает запись пачки сервером — только теперь она удаляется.
func (s *ResultStore) Confirm(batchId string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.inFlight == nil || s.inFlightId != batchId {
		return false
	}
	s.inFlight = nil
	s.inFlightId = ""
	s.dirty = true
	return true
}

// Clear полностью очищает накопленные результаты (очередь, неподтверждённую пачку
// и файл на диске). Используется сторожем связи при «удалении» пробы.
func (s *ResultStore) Clear() {
	s.mu.Lock()
	s.pending = nil
	s.inFlight = nil
	s.inFlightId = ""
	s.dirty = false
	s.mu.Unlock()
	_ = os.Remove(resultsFileName)
}

// Load восстанавливает недоставленные результаты после перезапуска.
func (s *ResultStore) Load() {
	data, err := os.ReadFile(resultsFileName)
	if err != nil {
		return // файла нет — чистый старт
	}
	var saved []ActionData
	if err := json.Unmarshal(data, &saved); err != nil {
		logResults.Error("Не удалось загрузить недоставленные результаты", "файл", resultsFileName, "ошибка", err)
		return
	}
	s.mu.Lock()
	s.pending = append(saved, s.pending...)
	s.mu.Unlock()
	if len(saved) > 0 {
		logResults.Info("Недоставленные результаты восстановлены с диска", "результатов", len(saved), "файл", resultsFileName)
		select {
		case s.signal <- struct{}{}:
		default:
		}
	}
}

// persistLoop периодически пишет снимок недоставленного на диск (только при изменениях).
//
// Probe:PersistIntervalSec = 0 означает «по таймеру не сохранять»: снимок делается
// только при остановке. Без этой проверки нулевой интервал ронял бы пробу при
// старте — таймер с неположительным периодом в Go паникует.
func (s *ResultStore) persistLoop() {
	if s.persist <= 0 {
		<-s.stop
		s.flush() // финальный снимок при остановке
		return
	}

	ticker := time.NewTicker(s.persist)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			s.flush()
		case <-s.stop:
			s.flush() // финальный снимок при остановке
			return
		}
	}
}

// flush атомарно записывает снимок «в полёте + очередь» через временный файл.
func (s *ResultStore) flush() {
	s.mu.Lock()
	if !s.dirty {
		s.mu.Unlock()
		return
	}
	s.dirty = false
	snapshot := slices.Concat(s.inFlight, s.pending) // slices.Concat — Go 1.22
	s.mu.Unlock()

	if len(snapshot) == 0 {
		_ = os.Remove(resultsFileName)
		return
	}
	data, err := json.Marshal(snapshot)
	if err != nil {
		logResults.Error("Не удалось сериализовать результаты", "ошибка", err)
		return
	}
	tmp := resultsFileName + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		logResults.Error("Не удалось сохранить результаты", "ошибка", err)
		return
	}
	if err := os.Rename(tmp, resultsFileName); err != nil {
		logResults.Error("Не удалось заменить файл результатов", "файл", resultsFileName, "ошибка", err)
	}
}

// Close останавливает фоновый цикл и делает финальный снимок.
func (s *ResultStore) Close() { close(s.stop) }

File: rust-probe/src/resultstore.rs
//! Хранилище результатов с гарантированной доставкой «ровно один раз» — аналог
//! Go ResultStore / C# ResultStore.
//!
//! Результаты копятся в очереди. TakeBatch выдаёт пачку «в полёте» с BatchId и ждёт
//! появления данных до таймаута (длинный опрос). Пачка не удаляется, пока сервер не
//! подтвердит её Confirm — значит, при любом сбое связи данные не теряются и не
//! задваиваются (сервер отбрасывает дубли по ResultId). Недоставленное переживает
//! перезапуск (снимок в файл results.json).

use crate::contracts::{new_guid, ActionData, ResultBatch, EMPTY_GUID};
use crate::logging::{f, Logger};
use std::sync::{Condvar, Mutex};
use std::time::Duration;

const RESULTS_FILE: &str = "results.json";
static LOG: Logger = Logger("results");

/// Внутреннее состояние под замком.
struct State {
    pending: Vec<ActionData>,
    in_flight: Option<Vec<ActionData>>,
    in_flight_id: String,
    dropped: u64,
    dirty: bool,
    stopped: bool,
}

/// Хранилище недоставленных результатов.
pub struct ResultStore {
    state: Mutex<State>,
    signal: Condvar,
    max_pending: usize,
}

impl ResultStore {
    pub fn new(max_pending: usize) -> Self {
        ResultStore {
            state: Mutex::new(State {
                pending: Vec::new(),
                in_flight: None,
                in_flight_id: String::new(),
                dropped: 0,
                dirty: false,
                stopped: false,
            }),
            signal: Condvar::new(),
            max_pending: max_pending.max(1),
        }
    }

    /// Добавляет результат в очередь; при переполнении вытесняет самый старый.
    pub fn add(&self, result: ActionData) {
        let mut st = self.state.lock().unwrap();
        st.pending.push(result);
        if st.pending.len() > self.max_pending {
            st.pending.remove(0);
            st.dropped += 1;
            if st.dropped == 1 || st.dropped % 1000 == 0 {
                LOG.error(
                    "Очередь результатов переполнена — старые записи вытесняются",
                    &[f("лимит", self.max_pending), f("всего_вытеснено", st.dropped)],
                );
            }
        }
        st.dirty = true;
        drop(st);
        self.signal.notify_all();
    }

    /// Выдаёт пачку результатов, ожидая появления данных до timeout (длинный опрос).
    /// Неподтверждённая пачка выдаётся повторно.
    pub fn take_batch(&self, timeout: Duration) -> ResultBatch {
        let mut st = self.state.lock().unwrap();
        loop {
            if let Some(items) = &st.in_flight {
                return ResultBatch { batch_id: st.in_flight_id.clone(), items: items.clone() };
            }
            if !st.pending.is_empty() {
                let items = std::mem::take(&mut st.pending);
                st.in_flight_id = new_guid();
                st.in_flight = Some(items.clone());
                return ResultBatch { batch_id: st.in_flight_id.clone(), items };
            }

            let (guard, wait) = self.signal.wait_timeout(st, timeout).unwrap();
            st = guard;
            if wait.timed_out() || st.stopped {
                return ResultBatch { batch_id: EMPTY_GUID.to_string(), items: Vec::new() };
            }
        }
    }

    /// Подтверждает запись пачки сервером — только теперь она удаляется.
    pub fn confirm(&self, batch_id: &str) -> bool {
        let mut st = self.state.lock().unwrap();
        if st.in_flight.is_none() || st.in_flight_id != batch_id {
            return false;
        }
        st.in_flight = None;
        st.in_flight_id.clear();
        st.dirty = true;
        true
    }

    /// Полностью очищает результаты (очередь, пачку «в полёте», файл на диске).
    /// Используется сторожем связи при «удалении» пробы.
    pub fn clear(&self) {
        let mut st = self.state.lock().unwrap();
        st.pending.clear();
        st.in_flight = None;
        st.in_flight_id.clear();
        st.dirty = false;
        drop(st);
        let _ = std::fs::remove_file(RESULTS_FILE);
    }

    /// Восстанавливает недоставленные результаты после перезапуска.
    pub fn load(&self) {
        let Ok(data) = std::fs::read(RESULTS_FILE) else {
            return; // файла нет — чистый старт
        };
        let saved: Vec<ActionData> = match serde_json::from_slice(&data) {
            Ok(v) => v,
            Err(e) => {
                LOG.error("Не удалось загрузить недоставленные результаты", &[f("файл", RESULTS_FILE), f("ошибка", e)]);
                return;
            }
        };
        if saved.is_empty() {
            return;
        }
        let count = saved.len();
        let mut st = self.state.lock().unwrap();
        let mut merged = saved;
        merged.append(&mut st.pending);
        st.pending = merged;
        drop(st);
        LOG.info("Недоставленные результаты восстановлены с диска", &[f("результатов", count), f("файл", RESULTS_FILE)]);
        self.signal.notify_all();
    }

    /// Атомарно пишет снимок «в полёте + очередь» на диск (только при изменениях).
    pub fn flush(&self) {
        let mut st = self.state.lock().unwrap();
        if !st.dirty {
            return;
        }
        st.dirty = false;
        let mut snapshot: Vec<ActionData> = st.in_flight.clone().unwrap_or_default();
        snapshot.extend(st.pending.iter().cloned());
        drop(st);

        if snapshot.is_empty() {
            let _ = std::fs::remove_file(RESULTS_FILE);
            return;
        }
        let data = match serde_json::to_vec(&snapshot) {
            Ok(d) => d,
            Err(e) => {
                LOG.error("Не удалось сериализовать результаты", &[f("ошибка", e)]);
                return;
            }
        };
        let tmp = format!("{RESULTS_FILE}.tmp");
        if let Err(e) = std::fs::write(&tmp, &data) {
            LOG.error("Не удалось сохранить результаты", &[f("ошибка", e)]);
            return;
        }
        if let Err(e) = std::fs::rename(&tmp, RESULTS_FILE) {
            LOG.error("Не удалось заменить файл результатов", &[f("файл", RESULTS_FILE), f("ошибка", e)]);
        }
    }

    /// Будит ожидающих на длинном опросе (используется при остановке службы).
    #[allow(dead_code)] // используется при штатной остановке службы
    pub fn wake(&self) {
        self.state.lock().unwrap().stopped = true;
        self.signal.notify_all();
    }
}
